File: bindings/python/irregex/compose.py
"""Narrow with exact match, then reason with compression (ADR-367).

`gist` answers *"where is this exact pattern?"*; `relate` answers *"what is this
text like?"*. These verbs need both at once: the exact `PatternSet` narrows the
corpus to a typed candidate set, then the compression kernel reasons **only
inside that narrowing**. Hand-piping `gist -l` into a relate verb throws the
match information away between the two steps and pays whole-corpus statistical
noise; these keep it.

  * `context(text, patterns)` — the minimal non-redundant reading set among files
    that *actually* match. `retrieval.pack` would happily rank a README that
    never mentions the subject.
  * `family(pattern)` — which implementations of this pattern are forks of each
    other, compared as **functions** rather than whole files, so one copy-pasted
    helper inside two unrelated modules still surfaces.
  * `provenance(text)` — where a pasted snippet came from, re-verified against the
    source file's current bytes rather than a shelf snapshot.

`radius.blast` is the fourth composed verb and lives in its own module.

`family`'s shape here is deliberately not the CLI's: a terminal reads top to
bottom, so it streams family and distinct rows interleaved. A program almost
always wants one class or the other, so it gets `FamilyReport.families` and
`.distinct` as separate fields.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

from . import engine
from .corpus import CORPUS_TIMEOUT, Region, Scope, merge_paths, region, run, scope_argv
from .grade import Channel
from .retrieval import Pick, decode_pick


if TYPE_CHECKING:
    from collections.abc import Sequence
    import os


@dataclass(frozen=True, slots=True)
class Family:
    """A verified family of implementations that match the same pattern. `edge` is the widest admitted distance in this family and `score` the engine's consolidation-opportunity ranking (repeated lines × channel confidence) — a display order, never a fused relevance."""

    rank: int
    unit: str
    channel: Channel
    edge: float
    repeated_lines: int
    score: float
    members: tuple[Region, ...]

    @property
    def size(self) -> int:
        """Members in this family."""
        return len(self.members)

    @property
    def paths(self) -> tuple[str, ...]:
        """Distinct files the family spans, in first-seen order."""
        return merge_paths(m.path for m in self.members)


@dataclass(frozen=True, slots=True)
class Distinct:
    """A matching implementation that joined no family — kept first-class, with its closest structural neighbor and both independent distances, so similar names never imply the same implementation by omission."""

    unit: str
    member: Region
    nearest: Region | None
    byte_distance: float
    structure_distance: float


@dataclass(frozen=True, slots=True)
class FamilyReport:
    """Both answer classes, split: `families` are consolidation candidates, `distinct` regions are the genuinely different implementations."""

    families: tuple[Family, ...]
    distinct: tuple[Distinct, ...]

    @property
    def paths(self) -> tuple[str, ...]:
        """Every file either class touches, families first."""
        return merge_paths(
            (p for f in self.families for p in f.paths),
            (d.member.path for d in self.distinct),
        )


@dataclass(frozen=True, slots=True)
class Attribution:
    """One phrase traced back to a file. `verified` is the whole point: the engine re-read `source`'s **current** bytes and re-found the phrase there, so `line` can be trusted. A `verified=False` row is drift — the shelf remembers the phrase, the live file no longer holds it."""

    text: str
    occurrences: int
    source: str
    verified: bool
    line: int | None

    def __str__(self) -> str:
        """`path:line` when verified, else a drift marker."""
        return f"{self.source}:{self.line}" if self.verified else f"(drift) {self.source or '?'}"


def _scoped(roots: Scope, corpus_wide: bool, verb: str) -> list[str]:
    """Lower the mandatory scope. `context`/`family` refuse to run unscoped so a composed query can never silently sweep `vendor/`; that refusal happens here rather than as an opaque exit 2 from the child."""
    args = scope_argv(roots)
    if corpus_wide:
        return ["--all", *args]
    if not args:
        msg = f"{verb} requires a scope: pass roots=… or corpus_wide=True"
        raise ValueError(msg)
    return args


def context(
    text: str,
    patterns: Sequence[str],
    *,
    match: Literal["any", "all"] = "any",
    fixed: bool = False,
    ignore_case: bool = False,
    top: int = 8,
    roots: Scope = (),
    corpus_wide: bool = False,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = CORPUS_TIMEOUT,
) -> list[Pick]:
    """The minimal non-redundant reading set among files that *actually* match `patterns`.

    `match="any"` admits a file that hit ≥1 pattern, `"all"` requires every one.
    Only the survivors are packed, so each `Pick` carries both its
    `marginal_bits` and the `patterns` that admitted it — two scores, never fused.
    """
    if not patterns:
        msg = "context: at least one pattern is required"
        raise ValueError(msg)
    argv = [text, *(flag for p in patterns for flag in ("-e", p)), "--match", match]
    if fixed:
        argv.append("-F")
    if ignore_case:
        argv.append("-i")
    argv += ["--top", str(top), *_scoped(roots, corpus_wide, "context")]
    rows, _ = run("irregex", "context", argv, cwd=cwd, timeout=timeout)
    return [decode_pick(r) for r in rows]


def family(
    pattern: str,
    *,
    unit: Literal["function", "match", "file"] = "function",
    max_structure_distance: float | None = None,
    max_distance: float | None = None,
    echo_min: float | None = None,
    context_lines: int | None = None,
    min_size: int = 2,
    fixed: bool = False,
    ignore_case: bool = False,
    top: int = 50,
    roots: Scope = (),
    corpus_wide: bool = False,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = CORPUS_TIMEOUT,
) -> FamilyReport:
    """Which implementations of `pattern` are forks of each other, compared as code rather than as files.

    `kinship.dups` compares whole files, so two unrelated modules that happen to
    share one copy-pasted helper stay invisible. This lifts every exact hit to its
    enclosing function first (`unit="match"` for bounded windows, `"file"` for
    whole-file kinship), then builds families only among those units.

    The three thresholds select the channel, mirroring the CLI's deliberately
    distinct spellings so a polarity can never silently invert:
    `max_structure_distance` for shared skeletons (the default), `max_distance`
    for byte near-duplicates, `echo_min` for renamed twins. Test files sharing a
    skeleton are structural twins rather than byte duplicates, so prefer
    `echo_min`/`max_structure_distance` when hunting those.
    """
    argv: list[str] = [pattern, "--unit", unit, "--min-size", str(min_size)]
    for flag, value in (
        ("--max-structure-distance", max_structure_distance),
        ("--max-distance", max_distance),
        ("--echo-min", echo_min),
        ("-C", context_lines),
    ):
        if value is not None:
            argv += [flag, str(value)]
    if fixed:
        argv.append("-F")
    if ignore_case:
        argv.append("-i")
    argv += ["--top", str(top), *_scoped(roots, corpus_wide, "family")]
    rows, _ = run("irregex", "family", argv, cwd=cwd, timeout=timeout)
    return FamilyReport(
        families=tuple(_family(r) for r in rows if r.get("kind") == "family"),
        distinct=tuple(_distinct(r) for r in rows if r.get("kind") == "distinct"),
    )


def _family(row: dict[str, object]) -> Family:
    """Decode one family row."""
    return Family(
        rank=engine.as_int(row, "rank"),
        unit=engine.as_str(row, "unit"),
        channel=Channel.parse(engine.as_str(row, "channel", "shapes")),
        edge=engine.as_float(row, "edge", 0.0) or 0.0,
        repeated_lines=engine.as_int(row, "repeated_lines"),
        score=engine.as_float(row, "score", 0.0) or 0.0,
        members=tuple(region(m) for m in engine.as_rows(row, "members")),
    )


def _distinct(row: dict[str, object]) -> Distinct:
    """Decode one unaffiliated-implementation row."""
    member = row.get("member")
    nearest = row.get("nearest")
    return Distinct(
        unit=engine.as_str(row, "unit"),
        member=region(member if isinstance(member, dict) else {}),
        nearest=region(nearest) if isinstance(nearest, dict) else None,
        byte_distance=engine.as_float(row, "byte_distance", 1.0) or 0.0,
        structure_distance=engine.as_float(row, "structure_distance", 1.0) or 0.0,
    )


def provenance(
    text: str,
    *,
    min_phrase: int = 12,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = CORPUS_TIMEOUT,
) -> list[Attribution]:
    """Where `text` came from, re-verified against the tree's current bytes.

    `retrieval.quote` attributes phrases from the codex shelf, which is a snapshot
    and can therefore point at a line that no longer exists. This re-reads each
    attributed file and re-finds the phrase, so a `verified` row is a citation you
    can follow. Corpus-wide by design — provenance reads the whole shelf, which
    `introspection.atlas_index(shelf=True)` builds.
    """
    rows, _ = run(
        "irregex",
        "provenance",
        [text, "--min-phrase", str(min_phrase)],
        cwd=cwd,
        timeout=timeout,
    )
    return [
        Attribution(
            text=engine.as_str(r, "text"),
            occurrences=engine.as_int(r, "occurrences"),
            source=engine.as_str(r, "source"),
            verified=bool(r.get("verified")),
            line=engine.as_int(r, "line") if r.get("line") is not None else None,
        )
        for r in rows
    ]

File: bindings/go/compose/compose.go
// Package compose is the plane where both engines answer one question (ADR-367):
// the exact pattern set narrows the corpus to a typed candidate set, and the
// compression kernel then reasons only INSIDE that subset.
//
// That containment is the point. A hand-rolled `gist -l | relate …` pipe throws
// the match information away between the two steps and then pays whole-corpus
// statistical noise; here the exact and statistical scores stay in separate
// fields and are never fused into one number.
//
// Scope is mandatory for [Corpus.Context] and [Corpus.Family] — a composed query
// must not silently sweep vendored trees. [Corpus.Blast] is corpus-wide by
// design, since a blast radius that stops at a directory boundary is a lie.
package compose

import (
	"context"
	"errors"

	irregex "irregex/bindings/go"
	"irregex/bindings/go/relate"
	"irregex/bindings/go/runtime"
)

// ErrUnscoped reports a composed query with neither roots nor All.
var ErrUnscoped = errors.New("irregex: composed query needs roots or Over().All()")

// Corpus is the scope a composed verb runs inside.
type Corpus struct {
	roots []string
	dir   string
	all   bool
}

// Over scopes the composed verbs to roots.
func Over(roots ...string) *Corpus { return &Corpus{roots: roots} }

// All declares the deliberate whole-corpus sweep, the only way to run a composed
// verb without roots.
func (c *Corpus) All() *Corpus {
	c.all = true
	return c
}

// In sets the working directory the query resolves relative paths against.
func (c *Corpus) In(dir string) *Corpus {
	c.dir = dir
	return c
}

func (c *Corpus) rows(ctx context.Context, op irregex.Op, p irregex.Compose, scoped bool) (*runtime.Rows, error) {
	if scoped && len(c.roots) == 0 && !c.all {
		return nil, ErrUnscoped
	}
	return runtime.Run(ctx, runtime.Query{Op: op, Params: p, Roots: c.roots, Dir: c.dir})
}

// Context is the reading set among the files that actually match — coverage
// packing restricted to the candidate set, so each pick reports both the patterns
// that admitted it and the novelty it adds over the picks before it.
func (c *Corpus) Context(ctx context.Context, p irregex.Compose) ([]relate.Pick, error) {
	rows, err := c.rows(ctx, irregex.OpContext, p, true)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []relate.Pick
	for row, err := range rows.All() {
		if err != nil {
			return out, err
		}
		out = append(out, relate.ScanPick(row))
	}
	return out, nil
}

// Family finds the fork families among only the matching files. Set
// Compose.MaxDistance for byte copy-paste; set Compose.MinEcho for the same
// skeleton under renamed vocabulary — which is what test families usually are, so
// a byte threshold there finds no edges at all.
func (c *Corpus) Family(ctx context.Context, p irregex.Compose) ([]relate.Family, error) {
	rows, err := c.rows(ctx, irregex.OpFamily, p, true)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []relate.Family
	for row, err := range rows.All() {
		if err != nil {
			return out, err
		}
		out = append(out, relate.ScanFamily(row))
	}
	return out, nil
}

// Attribution is one phrase of a pasted text traced to its source. Verified is
// the composed part: the exact engine re-checked the phrase against the source
// file's CURRENT bytes, so a stale statistical attribution cannot pass as live
// provenance.
type Attribution struct {
	Text        string
	Occurrences int64
	Source      string
	Verified    bool
	Line        int64
	Located     bool
}

// Provenance traces a pasted text back to the files it came from, each phrase
// re-verified against current bytes. Needs the codex shelf
// (`relate index --shelf`).
func (c *Corpus) Provenance(ctx context.Context, p irregex.Compose) ([]Attribution, error) {
	rows, err := c.rows(ctx, irregex.OpProvenance, p, false)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Attribution
	for row, err := range rows.All() {
		if err != nil {
			return out, err
		}
		line, located := row.OptInt("line")
		out = append(out, Attribution{
			Text:        row.Text("text"),
			Occurrences: row.Int("occurrences"),
			Source:      row.Text("source"),
			Verified:    row.Bool("verified"),
			Line:        line,
			Located:     located,
		})
	}
	return out, nil
}

// Reference is one place that mentions a symbol. Defines separates the definition
// from the two hundred call sites, and Enclosing names the function it sits in.
type Reference struct {
	Path      string
	Line      int64
	Enclosing string
	Defines   bool
}

// Dependency is a symbol the seed's own body depends on.
type Dependency struct {
	Symbol string
	Path   string
	Line   int64
}

// Ripple is a file reached indirectly, through Via, in Hops steps.
type Ripple struct {
	Path string
	Via  string
	Hops int64
}

// Mention is prose — a comment or doc line — that names the symbol. Renaming
// without these leaves the comments lying.
type Mention struct {
	Path string
	Line int64
	Text string
}

// Site is a definition's location.
type Site struct {
	Path string
	Line int64
}

// Blast is the live blast radius of one symbol, computed from CURRENT bytes with
// no precomputed graph, which is what makes it safe to consult mid-edit. Omitted
// is what Compose.Budget trimmed from the low-priority tail, so a bounded answer
// admits it was bounded.
type Blast struct {
	Symbol       string
	Kind         string
	Definitions  []Site
	Dependents   []Reference
	Dependencies []Dependency
	Twins        []relate.Neighbor
	Ripple       []Ripple
	Comments     []Mention
	Notes        []string
	Omitted      int64
}

// Blast answers "what moves if I change this symbol" before the edit: the
// definition and its kind, the direct dependents split into definitions and uses,
// what the body itself depends on, tangential twins, the same-language ripple, and
// the comments that name it.
func (c *Corpus) Blast(ctx context.Context, p irregex.Compose) (Blast, error) {
	rows, err := c.rows(ctx, irregex.OpBlast, p, false)
	if err != nil {
		return Blast{}, err
	}
	defer rows.Close()
	if !rows.Next() {
		return Blast{}, rows.Err()
	}
	row := rows.Row()
	b := Blast{
		Symbol:  row.Text("symbol"),
		Kind:    row.Text("kind"),
		Twins:   relate.Neighbors(row.Rows("twins")),
		Notes:   row.Strings("notes"),
		Omitted: row.Int("omitted"),
	}
	for _, d := range row.Rows("definitions") {
		b.Definitions = append(b.Definitions, Site{Path: d.Text("path"), Line: d.Int("line")})
	}
	for _, d := range row.Rows("dependents") {
		b.Dependents = append(b.Dependents, Reference{
			Path:      d.Text("path"),
			Line:      d.Int("line"),
			Enclosing: d.Text("enclosing"),
			Defines:   d.Bool("defines"),
		})
	}
	for _, d := range row.Rows("dependencies") {
		b.Dependencies = append(b.Dependencies, Dependency{Symbol: d.Text("symbol"), Path: d.Text("path"), Line: d.Int("line")})
	}
	for _, d := range row.Rows("ripple") {
		b.Ripple = append(b.Ripple, Ripple{Path: d.Text("path"), Via: d.Text("via"), Hops: d.Int("hops")})
	}
	for _, d := range row.Rows("comments") {
		b.Comments = append(b.Comments, Mention{Path: d.Text("path"), Line: d.Int("line"), Text: d.Text("text")})
	}
	return b, rows.Err()
}
